src/java/org/apache/cassandra/db/tries/MemtableReadTrie.java

```diff
@@ -0,0 +1,921 @@
+     instead of creating a new one. When a chain node already has a child and needs a new one added we change the type
+     (i.e. create a new node and remap the parent) to sparse with two children. When a six-child sparse node needs a new
+     child, we switch to split.
+
+     Blocks currently are not reused, because we do not yet have a mechanism to tell when readers are done with blocks
+     they are referencing. This currently causes a very low overhead (because we change data in place with the only
+     exception of nodes needing to change type) and is planned to be addressed later.
+
+     For further descriptions and examples of the mechanics of the trie, see MemtableTrie.md.
+     */
+
+    static final int BLOCK_SIZE = 32;
+
+    // Biggest block offset that can contain a pointer.
+    static final int LAST_POINTER_OFFSET = BLOCK_SIZE - 4;
+
+    /*
+     Block offsets used to identify node types (by comparing them to the node 'pointer offset').
+     */
+
+    // split node (dense, 2-3-3 transitions), laid out as 4 pointers to "mid" block, with has 8 pointers to "tail" block,
+    // which has 8 pointers to children
+    static final int SPLIT_OFFSET = BLOCK_SIZE - 4;
+    // sparse node, unordered list of up to 6 transition, laid out as 6 transition pointers followed by 6 transition
+    // bytes. The last two bytes contain an ordering of the transitions (in base-6) which is used for iteration. On
+    // update the pointer is set last, i.e. during reads the node may show that a transition exists and list a character
+    // for it, but pointer may still be null.
+    static final int SPARSE_OFFSET = BLOCK_SIZE - 2;
+    // min and max offset for a chain node. A block of chain node is laid out as a pointer at LAST_POINTER_OFFSET,
+    // preceded by characters that lead to it. Thus a full chain block contains BLOCK_SIZE-4 transitions/chain nodes.
+    static final int CHAIN_MIN_OFFSET = 0;
+    static final int CHAIN_MAX_OFFSET = BLOCK_SIZE - 5;
+    // Prefix node, an intermediate node augmenting its child node with content.
+    static final int PREFIX_OFFSET = BLOCK_SIZE - 1;
+
+    /*
+     Offsets and values for navigating in a block for particular node type. Those offsets are 'from the node pointer'
+     (not the block start) and can be thus negative since node pointers points towards the end of blocks.
+     */
+
+    // Limit for the starting cell / sublevel (2 bits -> 4 pointers).
+    static final int SPLIT_START_LEVEL_LIMIT = 4;
+    // Limit for the other sublevels (3 bits -> 8 pointers).
+    static final int SPLIT_OTHER_LEVEL_LIMIT = 8;
+    // Bitshift between levels.
+    static final int SPLIT_LEVEL_SHIFT = 3;
+
+    static final int SPARSE_CHILD_COUNT = 6;
+    // Offset to the first child pointer of a spare node (laid out from the start of the block)
+    static final int SPARSE_CHILDREN_OFFSET = 0 - SPARSE_OFFSET;
```

Comment:
```suggestion
    static final int SPARSE_CHILDREN_OFFSET = -SPARSE_OFFSET;
```

Reply:
It is specifically `0` (the position of the first child pointer in the block) minus `SPARSE_OFFSET` (the pointer with which it is being accessed). Though the second version is the same value, it does not mean the same thing.